lib/spack/spack/util/environment.py

```diff
@@ -461,12 +486,50 @@ def is_unset(self, var_name):
         # unset
         return (type(var_updates[-1]) == UnsetEnv)
 
+    def blacklist(self, banned):
+        """
+        Remove all modifications to variable names in banned.
+        """
+        self.env_modifications = list(filter(
+            lambda x: x.name not in banned, self.env_modifications))
+
     def clear(self):
         """
         Clears the current list of modifications
         """
         self.env_modifications = []
 
+    def reversed(self):
+        """
+        Returns the EnvironmentModifications object that will reverse self
+
+        Only creates reversals for additions to the environment, as reversing
+        ``unset`` and ``remove_path`` modifications is impossible.
```

Comment:
If the `reverse` logic were migrated to the environment action objects, `Unset` could record the value of the variable at the time that it is unset. I think adding a TODO is appropriate since this is likely best effort anyhow

Reply:
The variable may already be unset.
```
$ spack env activate foo
...
$ despacktivate
```

At despacktivate time, the variable is already unset before the python process starts and the `EnvironmentModifications` object is created.

It would be possible to have environments write a file at activation time from which to restore the previous environment, but I don't think it's worth doing. This best-effort version emulates what the `module unload` function already does. I suspect that we won't see many (if any) cases in which the `setup_run_environment` method unsets a variable or removes a path.

Reply:
Oh riiight. Er I suppose it could be stored in an intermediate variable like `UNSET_<the original variable name>`. Or perhaps storing the settings in a file. In either case I don't think that needs to be done here. I do think though that it is possible to handle (just that it doesn't necessarily need to be handled in this PR).